src/persist-client/src/impl/state.rs

```diff
@@ -105,9 +105,7 @@ where
             return Ok(Ok(()));
         }
 
-        self.trace.push((keys.to_vec(), desc.clone()));
-        debug_assert_eq!(&self.upper(), desc.upper());
-
+        self.push_batch(keys, &desc);
```

Comment:
This in itself is already a nice simplification. Small thing, but ... :raised_hands: